File: frontend/src/types/competitor.ts
/**
 * 竞品监控列表数据模型
 * 供 API 层、stores、视图层共用
 */
export interface CompetitorItem {
  id: number;
  name: string;
  domain: string;
  logoUrl?: string;
  /** 分类标签，如 SaaS工具 / AI产品 */
  category: string;
  /** 分类标签样式类型（v1 只监控同类 SaaS / App，不含电商） */
  categoryType: "saas" | "ai" | "app" | "tool";
  /** 监控页面名称列表 */
  pages: string[];
  /** 折叠的额外页面数量 */
  extraPages: number;
  /** 监控频率，如 每天 */
  frequency: string;
  frequencyDesc: string;
  /** 最近抓取相对时间，如 2 小时前 */
  lastFetchAgo: string;
  /** 最近抓取绝对时间 */
  lastFetchTime: string;
  /** 预计下次抓取时间，如「明天 08:00」/「即将抓取」/「已暂停」 */
  nextCrawlAt: string;
  /** 监控开关：true=开启，false=暂停 */
  enabled: boolean;
  /** 展示用中文状态：监控中 / 已暂停 */
  statusLabel: string;
  statusType: "success" | "warning" | "info";
  statusDesc: string;
  /** 最近变化条数 */
  changes: number;
  /** 今日新增变化 */
  todayChanges: number;
  /** 该竞品下的监控源（详情/编辑用） */
  sources?: MonitorSourceItem[];
  /** 软删除标记：非空表示已移入回收站（ISO 时间字符串） */
  deletedAt?: string | null;
  /** 新增竞品时若命中回收站里的同竞品并恢复，则为 true */
  restored?: boolean;
}

/** v1 支持的数据源类型（与后端 SOURCE_TYPE_REGISTRY 对齐） */
export type SourceType =
  | "homepage"
  | "pricing"
  | "changelog"
  | "blog"
  | "docs"
  | "status"
  | "rss"
  | "app_store";

/** 后端返回的"可选监控页面"目录项 */
export interface SourceTypeOption {
  type: SourceType;
  label: string;
  render: "browser" | "http";
  defaultIntervalMinutes: number;
  llmHint: string;
}

/** 新增竞品时提交的单个监控源 */
export interface MonitorSourceInput {
  sourceType: SourceType;
  url?: string;
  name?: string;
  intervalMinutes?: number;
}

/** 后端返回的监控源实体 */
export interface MonitorSourceItem {
  id: number;
  competitorId: number;
  sourceType: SourceType;
  label: string;
  name: string;
  url: string;
  renderMode: "browser" | "http";
  intervalMinutes: number;
  enabled: boolean;
  lastStatus?: string | null;
  lastError?: string | null;
  failCount?: number;
  lastCrawledAt?: string | null;
  /** 是否被「连续失败自动停用」：前端据此展示提示与重新启用按钮 */
  autoDisabled?: boolean;
}

/** 单个监控源的抓取结果 */
export interface CrawlSourceResult {
  sourceId: number;
  sourceName: string;
  sourceType: SourceType;
  status: "success" | "failed" | "skipped";
  httpStatus?: number | null;
  /** 相比上次基准是否发生变化 */
  changed: boolean;
  /** 是否本次才首次抓取（建立基准，不算变化） */
  firstTime: boolean;
  /** 本次抓取是否因此产出了一条情报事件（变化且通过显著性过滤） */
  eventCreated: boolean;
  error?: string | null;
  durationMs: number;
}

/** 一次「立即抓取」的汇总结果 */
export interface CrawlResult {
  competitorId: number;
  total: number;
  succeeded: number;
  failed: number;
  changed: number;
  results: CrawlSourceResult[];
}

/** 新增竞品的请求体 */
export interface CompetitorCreatePayload {
  name: string;
  officialUrl: string;
  category?: string;
  sources?: MonitorSourceInput[];
}

/** 编辑竞品的请求体（PATCH，语义为局部更新，字段均可选） */
export type CompetitorUpdatePayload = Partial<CompetitorCreatePayload> & {
  /** 监控开关：active=开启监控，paused=暂停监控 */
  status?: "active" | "paused";
};

/** 自动发现：单个监控页类型的查找结果 */
export interface DiscoveredSource {
  sourceType: SourceType;
  label: string;
  /** 找到时为地址，未找到为 null */
  url: string | null;
  found: boolean;
  /** 来源：link=首页链接 / sitemap / common=常见路径兜底 */
  origin: string;
  httpStatus?: number | null;
}

/** 自动发现：一次「自动寻找页面」的整体结果 */
export interface DiscoverResult {
  officialUrl: string;
  /** 官网首页是否可达（不可达时无法自动发现） */
  homepageReachable: boolean;
  sources: DiscoveredSource[];
}

/** 智能预填：根据竞品名称推断的官网地址与分类 */
export interface SuggestResult {
  officialUrl: string | null;
  category: string | null;
  /** llm=AI 推断 / probe=域名探测 / none=没识别出来 */
  source: string;
  message: string;
}

File: backend/app/schemas/competitor.py
from datetime import datetime, timedelta, timezone

from pydantic import BaseModel, ConfigDict, Field, computed_field, field_validator
from pydantic.alias_generators import to_camel

from app.core.timeutil import format_time, humanize_ago, to_utc
from app.models.competitor import CompetitorStatus
from app.schemas.source import MonitorSourceCreate, MonitorSourceOut, normalize_url

# 列表里最多平铺展示几个页面名，多出来的折叠成 "+N"
_MAX_VISIBLE_PAGES = 3

# 频率换算：按"周 → 天 → 小时"从大到小试，能整除就用对应单位
_INTERVAL_UNITS = ((10080, "周"), (1440, "天"), (60, "小时"))


def _humanize_interval(minutes: int) -> str:
    """把分钟数说成人话，如 1440 → 每天。"""
    for unit_minutes, unit in _INTERVAL_UNITS:
        if minutes % unit_minutes == 0:
            count = minutes // unit_minutes
            return f"每{unit}" if count == 1 else f"每 {count} {unit}"
    return f"每 {minutes} 分钟"


def _failed_sources(sources: list[MonitorSourceOut]) -> list[MonitorSourceOut]:
    """最近一次抓取失败的监控源（含因连续失败被自动停用的，状态要如实反映）。"""
    return [s for s in sources if s.last_status == "failed"]


def _auto_disabled_sources(sources: list[MonitorSourceOut]) -> list[MonitorSourceOut]:
    """因连续失败被自动停用、当前仍处于禁用态的监控源。"""
    return [s for s in sources if s.auto_disabled]


def _next_crawl_at(sources: list[MonitorSourceOut]) -> str:
    """预计下次抓取时间：取所有**启用中**的源里最早到期的那个。

    口径与调度器一致（上次抓取时间 + 该源自己的间隔）。
    调度器每分钟扫一次，所以这里给的是"最早可能被抓"的时间点——
    界面上用「即将」而不是精确到秒，避免给出比实际更精确的暗示。
    """
    enabled = [s for s in sources if s.enabled]
    if not enabled:
        return "已暂停"

    now = datetime.now(timezone.utc)
    due_times: list[datetime] = []
    for source in enabled:
        last = to_utc(source.last_crawled_at)
        if last is None:
            return "即将抓取"  # 从未抓过 → 下一轮扫描就会抓
        due_times.append(last + timedelta(minutes=max(1, source.interval_minutes)))

    earliest = min(due_times)
    if earliest <= now:
        return "即将抓取"

    local = earliest.astimezone()
    delta_days = (local.date() - datetime.now().astimezone().date()).days
    if delta_days == 0:
        return f"今天 {local:%H:%M}"
    if delta_days == 1:
        return f"明天 {local:%H:%M}"
    return f"{local.month}月{local.day}日 {local:%H:%M}"


class CompetitorCreate(BaseModel):
    """新增竞品时传的字段。

    official_url 必填：竞品监控的核心就是"盯官网"，没有网址无从监控。
    sources 可选：不传时后端自动为官网首页建一个监控源，做到"填完即可开始监控"。
    """

    # 前端按 camelCase 提交（officialUrl / sources[].sourceType），同时兼容下划线写法
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    name: str = Field(..., min_length=1, max_length=100)
    official_url: str = Field(..., min_length=1, max_length=255)
    category: str | None = Field(default=None, max_length=50)
    sources: list[MonitorSourceCreate] = Field(default_factory=list)

    @field_validator("name")
    @classmethod
    def _strip_name(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("竞品名称不能为空")
        return v

    @field_validator("official_url")
    @classmethod
    def _normalize_official_url(cls, v: str) -> str:
        value = normalize_url(v)
        if not value:
            raise ValueError("官网地址不能为空")
        return value


class CompetitorUpdate(BaseModel):
    """修改竞品：所有字段都可省略，只改传了的那些。

    sources 是"整份替换"语义：传了就以它为准做增删改，不传则完全不动监控源。
    """

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    name: str | None = Field(default=None, min_length=1, max_length=100)
    official_url: str | None = Field(default=None, min_length=1, max_length=255)
    category: str | None = Field(default=None, max_length=50)
    status: CompetitorStatus | None = None
    sources: list[MonitorSourceCreate] | None = None

    @field_validator("official_url")
    @classmethod
    def _normalize_official_url(cls, v: str | None) -> str | None:
        if not v or not v.strip():
            return None
        return normalize_url(v)


class CompetitorOut(BaseModel):
    model_config = ConfigDict(from_attributes=True, alias_generator=to_camel, populate_by_name=True)

    id: int
    user_id: int
    name: str
    official_url: str | None = None
    category: str | None = None
    status: CompetitorStatus
    created_at: datetime

    # 该竞品下的监控源（selectin 预加载，直接可读）
    sources: list[MonitorSourceOut] = Field(default_factory=list)

    # 由列表接口聚合填充：
    changes: int = 0
    today_changes: int = 0

    # ---- 以下为前端展示用的派生字段，均由上面的原始字段/监控源算出 ----

    @computed_field
    @property
    def domain(self) -> str:
        return self.official_url or ""

    @computed_field
    @property
    def category_type(self) -> str:
        return "saas"  # 前端用于 CSS 类名（tag-saas）

    @computed_field
    @property
    def pages(self) -> list[str]:
        return [s.name for s in self.sources[:_MAX_VISIBLE_PAGES]]

    @computed_field
    @property
    def extra_pages(self) -> int:
        return max(0, len(self.sources) - _MAX_VISIBLE_PAGES)

    @computed_field
    @property
    def frequency(self) -> str:
        intervals = [s.interval_minutes for s in self.sources if s.enabled]
        return _humanize_interval(min(intervals)) if intervals else "未配置"

    @computed_field
    @property
    def frequency_desc(self) -> str:
        return "定时抓取" if self.sources else "暂无监控源"

    @computed_field
    @property
    def last_fetch_ago(self) -> str:
        times = [s.last_crawled_at for s in self.sources if s.last_crawled_at]
        return humanize_ago(max(times), empty="从未抓取") if times else "从未抓取"

    @computed_field
    @property
    def last_fetch_time(self) -> str:
        times = [s.last_crawled_at for s in self.sources if s.last_crawled_at]
        return format_time(max(times)) if times else ""

    @computed_field
    @property
    def next_crawl_at(self) -> str:
        """预计下次抓取时间（由定时调度器驱动，见里程碑 10）。"""
        return _next_crawl_at(self.sources)

    @computed_field
    @property
    def enabled(self) -> bool:
        return self.status == CompetitorStatus.ACTIVE

    @computed_field
    @property
    def status_label(self) -> str:
        if self.status != CompetitorStatus.ACTIVE:
            return "已暂停"
        return "监控异常" if _failed_sources(self.sources) else "监控中"

    @computed_field
    @property
    def status_type(self) -> str:
        if self.status != CompetitorStatus.ACTIVE:
            return "info"
        return "warning" if _failed_sources(self.sources) else "success"

    @computed_field
    @property
    def status_desc(self) -> str:
        if self.status != CompetitorStatus.ACTIVE:
            return "手动暂停"
        failed = _failed_sources(self.sources)
        if not failed:
            return "正常"
        desc = f"抓取失败 {sum(s.fail_count for s in failed)} 次"
        stopped = _auto_disabled_sources(self.sources)
        if stopped:
            desc += f"，{len(stopped)} 个页面已停用"
        return desc


    @computed_field
    @property
    def logo_url(self) -> str | None:
        """首选图标地址：竞品官网自身的 favicon。

        不用 Clearbit 之类的第三方 logo 服务——它们不稳定且部分已弃用，
        而且前端还有 favicon → apple-touch-icon → 首字母头像的多级回退。
        """
        if not self.official_url:
            return None
        host = self.official_url.split("//")[-1].split("/")[0].strip()
        host = host[4:] if host.startswith("www.") else host
        return f"https://{host}/favicon.ico" if host else None


class FaviconOut(BaseModel):
    """真实图标解析结果（给 SPA 站点兜底，见 services/favicon.py）。"""

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    logo_url: str | None = None


class SuggestRequest(BaseModel):
    """「智能预填」接口的请求：竞品名称 + 可选分类候选项。

    use_llm：是否允许调用 AI 推断官网/分类。
    - True（默认）：「智能检测填充」按钮走 LLM，能处理中文品牌名等复杂情况；
    - False：用户自己填竞品时的自动预填，只用规则域名探测，不消耗 AI。
    """

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    name: str = Field(..., min_length=1, max_length=100)
    categories: list[str] = Field(default_factory=list)
    use_llm: bool = True


class SuggestResult(BaseModel):
    """「智能预填」接口的响应：推断出的官网地址与分类（可能为空）。"""

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    official_url: str | None = None
    category: str | None = None
    source: str = "none"  # llm | probe | none
    message: str = ""
